content/en/blog/_posts/2021-09-07-data-duplication-in-data-heavy-k8s-env.md

```diff
@@ -0,0 +1,221 @@
+version in 1.20. Volume Snapshots basically allows Kubernetes to use the cloud-native storage 
+snapshotting mechanism that allows data copy to offload from compute to the actual storage device.
+
+Since data is already in the same storage device (array of devices), duplicating data is usually 
+a metadata operation. All you need to do is point a new disk to an immutable snapshot and only 
+save deltas (or let it do a full-disk copy). As an operation that is inside the storage back-end,
+it’s much quicker and usually doesn’t involve sending traffic over the network.
+
+## Sounds like we have our solution, right?
+
+Actually, Volume Snapshots are namespaced, and Kubernetes protects namespaced data from 
+being shared between tenants (namespaces). This Kubernetes limitation is a conscious design decision so that a pod in a different namespace can’t mount another application’s Persistent Volume (Claim).
+
+One way around it would be to create duplicate data accessible on the same namespace. 
+But if it says that it’s the “production” namespace and many teams are working on this data,
+someone with a fat finger might accidentally swap config files and end up messing up 
+production or data used by other teams.
+
+So the idea is to separate teams/initiatives by namespaces to avoid that and generally 
+limit access to the production namespace.
+
+## Solution? Creating a Golden Snapshot externally
+
+Another way around this design limitation is to create Snapshot externally (not through Kubernetes).
+This is also called pre-provisioning a snapshot manually. Next, I will import it 
+as a multi-tenant golden snapshot that can be used for many namespaces.
+
+High-level plan for preparing the Golden Snapshot:
+1. Identify Disk (EBS/Persistent Disk) that you want to clone with data in the cloud provider
+2. Make a Disk Snapshot (in CSP console)
+3. Get Disk Snapshot ID
+
+High-level plan for cloning data for each team
+1. Create namespace “sandbox01”
+2. Import Disk Snapshot (ID) as Volume Snapshot Content to Kubernetes
+3. Create Volume Snapshot in the namespace sandbox01 mapped to Volume Snapshot Content
+4. Create the Persistent Volume Claim from Volume Snapshot
+5. Install deployment/statefulSet with PVC
+
+## Step 1: Identify Disk
+
+First, you need to identify your golden source. In my case, it’s a PostgreSQL database
+on PersistentVolumeClaim “postgres-pv-claim” in the “production” namespace.
+
+```terminal
+kubectl -n production get pvc postgres-pv-claim -o jsonpath='{.spec.volumeName}'
+```
+
+The output will look like this: ```pvc-3096b3ba-38b6-4fd1-a42f-ec99176ed0d90```
+
+## Step 2: Prepare your golden source
+
+You need to do this once or every time you want to refresh your golden data.
+
+### Make a Disk Snapshot
+
+Go to AWS EC2 or GCP Compute Engine console and search for EBS/Persistent Disk 
+with the label ```pvc-3096b3ba-38b6-4fd1-a42f-ec99176ed0d9```.
```

Comment:
I think we should pick one cloud provider and consistently use it as an example. For example, instead of repeatedly saying "AWS or GCP", let's just stick with GCP. Perhaps add a note at the introduction that GCP is merely used as an illustrative example, and it's possible with other cloud providers.

Reply:
I think the core message of this blog post, that these features are not well documented and it's not clear how to create VolumeSnapshotContent and reference it in YAML as there is no consistency between CSI implementations.
For example:
   snapshotHandle: 'gcp/projects/staging-eu-castai-vt5hy2/global/snapshots/6645363163809389170'
   snapshotHandle: 'snap-07ff83d328c981c98'

So I'm leaning towards showing this inconsistency, but if it's a deal-breaker we could leave single cloud

Reply:
```suggestion
Go to AWS EC2 or GCP Compute Engine console and search for an EBS volume
(on AWS) or Persistent Disk (on GCP), that has a label matching the last output.
In this case I saw: `pvc-3096b3ba-38b6-4fd1-a42f-ec99176ed0d9`.
```